source/distributed/tria.cc

```diff
@@ -1124,10 +1124,545 @@ namespace
+      AssertThrowMPI(ierr);
+      // this barrier is necessary, because otherwise others might already
+      // write while one core is still setting the size to zero.
+      ierr = MPI_Barrier(mpi_communicator);
+      AssertThrowMPI(ierr);
+      ierr = MPI_Info_free(&info);
+      AssertThrowMPI(ierr);
+      // ------------------
+
+      // Check if number of processors is lined up with p4est partitioning.
+      Assert(myrank < parallel_forest->mpisize, ExcInternalError());
+
+      // Write cumulative sizes to file.
+      if (myrank == 0)
+        MPI_File_write_at(fh,
+                          0,
+                          cumulative_sizes_fix.data(),
+                          cumulative_sizes_fix.size(),
+                          MPI_UNSIGNED,
+                          MPI_STATUS_IGNORE);
+
+      // Write packed data to file simultaneously.
+      const unsigned int offset =
+        cumulative_sizes_fix.size() * sizeof(unsigned int);
+
+      MPI_File_write_at(
+        fh,
+        offset + parallel_forest->global_first_quadrant[myrank] *
+                   cumulative_sizes_fix.back(), // global position in file
+        src_data_fix.data(),
+        src_data_fix.size(), // local buffer
+        MPI_CHAR,
+        MPI_STATUS_IGNORE);
+
+      ierr = MPI_File_close(&fh);
+      AssertThrowMPI(ierr);
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::DataTransfer::load(
+      const typename dealii::internal::p4est::types<dim>::forest
+        *                parallel_forest,
+      const char *       filename,
+      const unsigned int n_attached_deserialize)
+    {
+      Assert(dest_data_fix.capacity() == 0,
+             ExcMessage("Previously loaded data has not been released yet!"));
+
+      const std::string fname_fixed = std::string(filename) + "_fixed.data";
+
+      // ----- copied -----
+      // from DataOutInterface::write_vtu_parallel
+      // TODO: write general MPIIO interface
+      int myrank, nproc;
+      int ierr = MPI_Comm_rank(mpi_communicator, &myrank);
+      AssertThrowMPI(ierr);
+      ierr = MPI_Comm_size(mpi_communicator, &nproc);
+      AssertThrowMPI(ierr);
+
+      MPI_Info info;
+      ierr = MPI_Info_create(&info);
+      AssertThrowMPI(ierr);
+      MPI_File fh;
+      ierr = MPI_File_open(mpi_communicator,
+                           const_cast<char *>(fname_fixed.c_str()),
+                           MPI_MODE_RDONLY,
+                           info,
+                           &fh);
+      AssertThrowMPI(ierr);
+
+      ierr = MPI_Info_free(&info);
+      AssertThrowMPI(ierr);
+      // ------------------
+
+      // Check if number of processors is lined up with p4est partitioning.
+      Assert(myrank < parallel_forest->mpisize, ExcInternalError());
+
+      // Read cumulative sizes from file.
+      cumulative_sizes_fix.resize(1 + n_attached_deserialize);
+      MPI_File_read_at(fh,
```

Comment:
say in the comment that *all* processors are reading this but will be reading the same information from the same location